test/functional/owners_controller_test.rb

```diff
@@ -0,0 +1,213 @@
+      sign_in_as(user)
+    end
+
+    context "on GET to index" do
+      context "when user owns the gem" do
+        setup do
+          get :index, params: { rubygem_id: @rubygem.name }
+        end
+
+        should respond_with :success
+        should "render all gem owners in owners table" do
+          @rubygem.ownerships.each do |o|
+            assert page.has_content?(o.owner_name)
+          end
+        end
+      end
+
+      context "when user does not own the gem" do
+        setup do
+          @other_user = create(:user)
+          sign_in_as(@other_user)
+          get :index, params: { rubygem_id: @rubygem.name }
+        end
+
+        should redirect_to("the sign in page") { sign_in_path }
+      end
+    end
+
+    context "on POST to create ownership" do
+      context "with correct params" do
+        setup do
+          @new_owner = create(:user)
+          post :create, params: { handle: @new_owner.display_id, rubygem_id: @rubygem.name }
+        end
+
+        should redirect_to("ownerships index") { rubygem_owners_path(@rubygem) }
+        should "add unconfirmed ownership record" do
+          assert @rubygem.owners_including_unconfirmed.include?(@new_owner)
+          assert_nil @rubygem.ownerships_including_unconfirmed.find_by(user: @new_owner).confirmed_at
+        end
+        should "set success notice flash" do
+          expected_notice = "Owner added successfully. A confirmation mail has been sent to #{@new_owner.handle}'s email"
+          assert_equal expected_notice, flash[:notice]
+        end
+      end
+
+      context "with incorrect params" do
+        should "show error message if user not found" do
+          post :create, params: { handle: "no_user", rubygem_id: @rubygem.name }
+          expected_alert = "User must exist"
+          assert_equal expected_alert, flash[:alert]
+        end
+
+        should "show error message if ownership exists" do
+          @new_owner = create(:user)
+          create(:ownership, rubygem: @rubygem, user: @new_owner)
+          post :create, params: { handle: @new_owner.handle, rubygem_id: @rubygem.name }
+          expected_alert = "User has already been taken"
+          assert_equal expected_alert, flash[:alert]
+        end
+      end
+    end
+
+    context "on DELETE to owners" do
+      setup do
+        @owner = create(:user)
+        @ownership = create(:ownership, rubygem: @rubygem, user: @owner)
+      end
+
+      context "remove user as gem owner" do
+        setup do
+          delete :destroy, params: { rubygem_id: @rubygem.name, handle: @owner.display_id }
+        end
+        should redirect_to("ownership index") { rubygem_owners_path(@rubygem) }
+        should "remove the ownership record" do
+          refute @rubygem.owners_including_unconfirmed.include?(@owner)
+        end
+      end
+
+      context "not remove last confirmed owner" do
+        setup do
+          @ownership.destroy
```

Comment:
IMHO, creating ownership and then destroying ownership in a nested setup feel unnecessary and makes the test hard to read. I don't think context at line 78 and 88 need a shared setup.
Note that, readability is more important in tests than code duplication.